[JSC] Use more `enum class` for Safer C++

Gumichocopengin8 opened the pull request:
#### e3cf8388bf12f51e3a7a7fafb6e4f3e3c419691a
<pre>
[JSC] Use more `enum class` for Safer C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=295733">https://bugs.webkit.org/show_bug.cgi?id=295733</a>

Reviewed by Darin Adler and Sosuke Suzuki.

This is a refactoring-only patch with no behavior changes.
Use `enum class` for `Kind`, `ArgumentRequirement`, `ConstantMode`, `DateTimeFormat`, `Type`, `SourceTextWhereErrorOccurred`,
`StackCaptureAction`, `CompilationKind`, `EmptyIdentifierFlag` for Safer C++ [1].

[1]: <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#use-enum-classes-instead-of-old-style-enumerations">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#use-enum-classes-instead-of-old-style-enumerations</a>

* Source/JavaScriptCore/API/OpaqueJSString.cpp:
(OpaqueJSString::identifier const):
* Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
(JSC::tryAllocate):
* Source/JavaScriptCore/runtime/BufferMemoryHandle.cpp:
(JSC::BufferMemoryManager::tryAllocateFastMemory):
(JSC::BufferMemoryManager::tryAllocateGrowableBoundsCheckingMemory):
(JSC::BufferMemoryManager::tryAllocatePhysicalBytes):
* Source/JavaScriptCore/runtime/BufferMemoryHandle.h:
* Source/JavaScriptCore/runtime/CommonIdentifiers.cpp:
(JSC::CommonIdentifiers::CommonIdentifiers):
* Source/JavaScriptCore/runtime/ConsoleClient.cpp:
(JSC::ConsoleClient::internalMessageWithTypeAndLevel):
(JSC::ConsoleClient::logWithLevel):
(JSC::ConsoleClient::clear):
(JSC::ConsoleClient::dir):
(JSC::ConsoleClient::dirXML):
(JSC::ConsoleClient::table):
(JSC::ConsoleClient::trace):
(JSC::ConsoleClient::assertion):
(JSC::ConsoleClient::group):
(JSC::ConsoleClient::groupCollapsed):
(JSC::ConsoleClient::groupEnd):
* Source/JavaScriptCore/runtime/ConsoleClient.h:
(): Deleted.
* Source/JavaScriptCore/runtime/ConstantMode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/ConstantMode.h:
(JSC::modeForIsConstant):
(): Deleted.
* Source/JavaScriptCore/runtime/DateConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/DateConversion.cpp:
(JSC::formatDateTime):
* Source/JavaScriptCore/runtime/DateConversion.h:
(): Deleted.
* Source/JavaScriptCore/runtime/DatePrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/DeletePropertySlot.h:
(JSC::DeletePropertySlot::DeletePropertySlot):
(JSC::DeletePropertySlot::setConfigurableMiss):
(JSC::DeletePropertySlot::setNonconfigurable):
(JSC::DeletePropertySlot::setHit):
(JSC::DeletePropertySlot::isCacheableDelete const):
(JSC::DeletePropertySlot::isDeleteHit const):
(JSC::DeletePropertySlot::isConfigurableDeleteMiss const):
(JSC::DeletePropertySlot::isNonconfigurable const):
(): Deleted.
* Source/JavaScriptCore/runtime/ErrorInstance.cpp:
(JSC::appendSourceToErrorMessage):
* Source/JavaScriptCore/runtime/ErrorInstance.h:
* Source/JavaScriptCore/runtime/Exception.h:
* Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
(JSC::defaultSourceAppender):
(JSC::notAFunctionSourceAppender):
(JSC::invalidParameterInSourceAppender):
(JSC::invalidParameterInstanceofSourceAppender):
(JSC::invalidPrototypeSourceAppender):
* Source/JavaScriptCore/runtime/ExecutableBase.h:
(): Deleted.
* Source/JavaScriptCore/runtime/Identifier.h:
(): Deleted.
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::ensureTerminationException):
* Source/JavaScriptCore/wasm/WasmMemory.cpp:
* Source/WebCore/bindings/js/JSDOMExceptionHandling.cpp:
(WebCore::reportException):

Canonical link: <a href="https://commits.webkit.org/302621@main">https://commits.webkit.org/302621@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e7acd58fd2f5f40fffd07135362e842a8f7d7c41

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129616 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1873 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40455 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137003 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81063 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131487 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1806 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1749 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98739 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66588 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132563 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1406 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116099 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79403 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1324 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34230 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80276 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/121608 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109798 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34728 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139482 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/128068 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1663 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1586 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107260 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1705 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112440 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107106 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27287 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1362 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30952 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/54415 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1734 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65097 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/161082 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1554 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40176 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1588 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1656 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/ConsoleClient.h:
```diff
@@ -80,7 +80,7 @@ class ConsoleClient : public CanMakeWeakPtr<ConsoleClient> {
 
 private:
-    enum ArgumentRequirement { ArgumentRequired, ArgumentNotRequired };
+    enum class ArgumentRequirement { ArgumentRequired, ArgumentNotRequired };
```
In a case like this we should change the enumeration values since they are repetitive once we switch to an `enum class`.
```suggestion
    enum class ArgumentRequirement { Required, NotRequired };
```
Or possibly even shorter, something like this.
```suggestion
    enum class ArgumentRequirement { No, Yes };
```

Gumichocopengin8 replied:
Fixed with `Yes` and `No`.

darinadler commented on Source/JavaScriptCore/runtime/DateConversion.h:
```diff
@@ -31,7 +31,7 @@ namespace JSC {
+enum class DateTimeFormat {
     DateTimeFormatDate = 1,
     DateTimeFormatTime = 2,
     DateTimeFormatDateAndTime = DateTimeFormatDate | DateTimeFormatTime
```
```suggestion
enum class DateTimeFormat {
    Date = 1,
    Time = 2,
    DateAndTime = Date | Time
```

Gumichocopengin8 replied:
Fixed.

darinadler reviewed (approved):
Please consider for each of these whether the enumeration values should be shortened since the values are now always in a named scope. I don't think it's good to leave the names unnecessarily long and repetitive. I'm approving without requiring changes but I strongly request you make that change

sosukesuzuki reviewed (approved):
r=me

Gumichocopengin8 committed:
[JSC] Use more `enum class` for Safer C++
https://bugs.webkit.org/show_bug.cgi?id=295733

Reviewed by Darin Adler and Sosuke Suzuki.

This is a refactoring-only patch with no behavior changes.
Use `enum class` for `Kind`, `ArgumentRequirement`, `ConstantMode`, `DateTimeFormat`, `Type`, `SourceTextWhereErrorOccurred`,
`StackCaptureAction`, `CompilationKind`, `EmptyIdentifierFlag` for Safer C++ [1].

[1]: https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#use-enum-classes-instead-of-old-style-enumerations

* Source/JavaScriptCore/API/OpaqueJSString.cpp:
(OpaqueJSString::identifier const):
* Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
(JSC::tryAllocate):
* Source/JavaScriptCore/runtime/BufferMemoryHandle.cpp:
(JSC::BufferMemoryManager::tryAllocateFastMemory):
(JSC::BufferMemoryManager::tryAllocateGrowableBoundsCheckingMemory):
(JSC::BufferMemoryManager::tryAllocatePhysicalBytes):
* Source/JavaScriptCore/runtime/BufferMemoryHandle.h:
* Source/JavaScriptCore/runtime/CommonIdentifiers.cpp:
(JSC::CommonIdentifiers::CommonIdentifiers):
* Source/JavaScriptCore/runtime/ConsoleClient.cpp:
(JSC::ConsoleClient::internalMessageWithTypeAndLevel):
(JSC::ConsoleClient::logWithLevel):
(JSC::ConsoleClient::clear):
(JSC::ConsoleClient::dir):
(JSC::ConsoleClient::dirXML):
(JSC::ConsoleClient::table):
(JSC::ConsoleClient::trace):
(JSC::ConsoleClient::assertion):
(JSC::ConsoleClient::group):
(JSC::ConsoleClient::groupCollapsed):
(JSC::ConsoleClient::groupEnd):
* Source/JavaScriptCore/runtime/ConsoleClient.h:
(): Deleted.
* Source/JavaScriptCore/runtime/ConstantMode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/ConstantMode.h:
(JSC::modeForIsConstant):
(): Deleted.
* Source/JavaScriptCore/runtime/DateConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/DateConversion.cpp:
(JSC::formatDateTime):
* Source/JavaScriptCore/runtime/DateConversion.h:
(): Deleted.
* Source/JavaScriptCore/runtime/DatePrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/DeletePropertySlot.h:
(JSC::DeletePropertySlot::DeletePropertySlot):
(JSC::DeletePropertySlot::setConfigurableMiss):
(JSC::DeletePropertySlot::setNonconfigurable):
(JSC::DeletePropertySlot::setHit):
(JSC::DeletePropertySlot::isCacheableDelete const):
(JSC::DeletePropertySlot::isDeleteHit const):
(JSC::DeletePropertySlot::isConfigurableDeleteMiss const):
(JSC::DeletePropertySlot::isNonconfigurable const):
(): Deleted.
* Source/JavaScriptCore/runtime/ErrorInstance.cpp:
(JSC::appendSourceToErrorMessage):
* Source/JavaScriptCore/runtime/ErrorInstance.h:
* Source/JavaScriptCore/runtime/Exception.h:
* Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
(JSC::defaultSourceAppender):
(JSC::notAFunctionSourceAppender):
(JSC::invalidParameterInSourceAppender):
(JSC::invalidParameterInstanceofSourceAppender):
(JSC::invalidPrototypeSourceAppender):
* Source/JavaScriptCore/runtime/ExecutableBase.h:
(): Deleted.
* Source/JavaScriptCore/runtime/Identifier.h:
(): Deleted.
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::ensureTerminationException):
* Source/JavaScriptCore/wasm/WasmMemory.cpp:
* Source/WebCore/bindings/js/JSDOMExceptionHandling.cpp:
(WebCore::reportException):

Canonical link: https://commits.webkit.org/302621@main
